### ISSUE TITLE
CNREL (mobile) login request not handled correctly

### ISSUE THREAD

**Issue by MlgmXyysd** [Melledy/Grasscutter#51]
When I try to login in `CNRELAndroid2.6.0` and `CNRELiOS2.6.0`, the game prompts a network error.

I have:
- [x] Connect to mitmproxy using WiFi proxy or VPN
- [x] Installed mitmproxy CA cert
- [x] Trusted mitmproxy CA cert

It seems that the server did not process the login request correctly for the following domains
```
gameapi-account.mihoyo.com
minor-api.mihoyo.com
public-data-api.mihoyo.com
uspider.yuanshen.com
sdk-static.mihoyo.com
```

**Comment by 4Benj** [comment]
Can you tell if the network request is going to grasscutter or the official servers?

**Comment by MlgmXyysd** [comment]
> Can you tell if the network request is going to grasscutter or the official servers?

I think all requests are redirected to Grasscutter https://github.com/Grasscutters/Grasscutter/blob/stable/proxy.py#L54

**Comment by 4Benj** [comment]
Add this `Grasscutter.getLogger().info(body);` at around 215 in emu.grasscutter.server.dispatch.DispatchServer
You should see what information is being sent to the server (If it's being sent at all).

**Comment by memetrollsXD** [comment]
CN uses slightly different subdomains

**Comment by MlgmXyysd** [comment]
> Add this `Grasscutter.getLogger().info(body);` at around 215 in emu.grasscutter.server.dispatch.DispatchServer You should see what information is being sent to the server (If it's being sent at all).

Now I can see the login request and logged in, so the problem turned to `4206` error code.

**Comment by MlgmXyysd** [comment]
![image](https://user-images.githubusercontent.com/17919858/164247436-669419da-6188-4ad3-b897-f809e043adf2.png)
PC (local) works well, iOS (LAN with WiFi proxy) not work

**Comment by KingRainbow44** [comment]
make sure your dispatch server ip and game server IP arent `127.0.0.1` but are your computer's local IP (usually like `192.168.1.x`)

**Comment by MlgmXyysd** [comment]
> make sure your dispatch server ip and game server IP arent `127.0.0.1` but are your computer's local IP (usually like `192.168.1.x`)

That's unlikely. But I can connect to the remote server on PC client, why?

**Comment by KingRainbow44** [comment]
assuming you have the downloading resource error, this comes from the client not being able to ping the **game server** (the server on UDP). this is because the game server IP in the grasscutter config is usually `127.0.0.1`. the IP you provide in the config for the game server is what is sent in the region query. if the IP in the region query is `127.0.0.1`, external devices cannot connect.

**Comment by KingRainbow44** [comment]
this is quite the common error, so ill be adding this to the troubleshooting wiki soon
also let me know if this fixes your issue

**Comment by MlgmXyysd** [comment]
> assuming you have the downloading resource error, this comes from the client not being able to ping the **game server** (the server on UDP). this is because the game server IP in the grasscutter config is usually `127.0.0.1`. the IP you provide in the config for the game server is what is sent in the region query. if the IP in the region query is `127.0.0.1`, external devices cannot connect.

Got it now, I need to change the address to a LAN accessible one in config.json, right?

**Comment by KingRainbow44** [comment]
yep

**Comment by MlgmXyysd** [comment]
> yep

One more question, what if I want to open it on the Public Internet (WAN)?

**Comment by MlgmXyysd** [comment]
> yep

sadly, tried both `GameServerIp` & `GameServerPublicIp`, still not work
![IMG_20220421_010338__01](https://user-images.githubusercontent.com/17919858/164285464-f0bb995c-c731-4cb0-8c2b-1899a5f36df5.jpg)
![IMG_20220421_010620](https://user-images.githubusercontent.com/17919858/164285475-2881be3a-b073-443b-b345-0205911e9958.jpg)

**Comment by lilmayofuksu** [comment]
listening on 127.0.0.1 (gameserverip) will probably reject any connections other than the localhost
try setting both of the ip sections to your lan ip

**Comment by lilmayofuksu** [comment]
> > yep
> 
> sadly, tried both `GameServerIp` & `GameServerPublicIp`, still not work ![IMG_20220421_010338__01](https://user-images.githubusercontent.com/17919858/164285464-f0bb995c-c731-4cb0-8c2b-1899a5f36df5.jpg) ![IMG_20220421_010620](https://user-images.githubusercontent.com/17919858/164285475-2881be3a-b073-443b-b345-0205911e9958.jpg)

also this is extremely weird, i'll have a look at this

**Comment by MlgmXyysd** [comment]
> listening on 127.0.0.1 (gameserverip) will probably reject any connections other than the localhost try setting both of the ip sections to your lan ip

already tried

**Comment by lilmayofuksu** [comment]
> > yep
> 
> sadly, tried both `GameServerIp` & `GameServerPublicIp`, still not work ![IMG_20220421_010338__01](https://user-images.githubusercontent.com/17919858/164285464-f0bb995c-c731-4cb0-8c2b-1899a5f36df5.jpg) ![IMG_20220421_010620](https://user-images.githubusercontent.com/17919858/164285475-2881be3a-b073-443b-b345-0205911e9958.jpg)

can you also show your dispatch ip's in this scenario?

**Comment by MlgmXyysd** [comment]
> can you also show your dispatch ip's in this scenario?

1. DispatchIp: "127.0.0.1" & DispatchPublicIp: "" & GameServerIp: "127.0.0.1" & GameServerPublicIp: "" -> 4206
2. DispatchIp: "127.0.0.1" & DispatchPublicIp: "" & GameServerIp: "192.168.1.100" & GameServerPublicIp: "" -> 4206
3. DispatchIp: "127.0.0.1" & DispatchPublicIp: "" & GameServerIp: "127.0.0.1" & GameServerPublicIp: "192.168.1.100" -> 4206
4. DispatchIp: "127.0.0.1" & DispatchPublicIp: "192.168.1.100" & GameServerIp: "127.0.0.1" & GameServerPublicIp: "" -> 4206
5. DispatchIp: "127.0.0.1" & DispatchPublicIp: "192.168.1.100" & GameServerIp: "127.0.0.1" & GameServerPublicIp: "192.168.1.100" -> 4206
6. DispatchIp: "127.0.0.1" & DispatchPublicIp: "192.168.1.100" & GameServerIp: "192.168.1.100" & GameServerPublicIp: "" -> 4206
7. DispatchIp: "192.168.1.100" -> mitmproxy error: 127.0.0.1:443 server refused

**Comment by lilmayofuksu** [comment]
DispatchIp: "192.168.1.100" & GameServerIp: "192.168.1.100" with public ip's empty, have you tried this?

**Comment by MlgmXyysd** [comment]
Proxy will always error if DispatchIp is not 127.0.0.1

**Comment by lilmayofuksu** [comment]
i think the proxy errors out because genshin tries to query the dispatch ip that it receives, and in that case that would be 127.0.0.1

**Comment by lilmayofuksu** [comment]
if you listen on 127.0.0.1 (on ip's that are not public), then no other device can connect to your server

**Comment by lilmayofuksu** [comment]
the public ip section just changes the ip's that are sent to the client

**Comment by MlgmXyysd** [comment]
> if you listen on 127.0.0.1 (on ip's that are not public), then no other device can connect to your server

but if I listen on other ip, localhost can't connect🤔

**Comment by lilmayofuksu** [comment]
that's also weird, this whole thing is very weird

**Comment by MlgmXyysd** [comment]
Fixed #78
